PayPalDataCollector/src/main/java/com/paypal/android/paypaldatacollector/PayPalDataCollector.kt

```diff
@@ -0,0 +1,73 @@
+package com.paypal.android.paypaldatacollector
+
+import android.content.Context
+import android.util.Log
+import com.paypal.android.core.CoreConfig
+import com.paypal.android.core.Environment
+import lib.android.paypal.com.magnessdk.InvalidInputException
+import lib.android.paypal.com.magnessdk.MagnesSDK
+import lib.android.paypal.com.magnessdk.MagnesSettings
+import lib.android.paypal.com.magnessdk.MagnesSource
+
+/**
+ * Enables you to collect data about a customer's device and correlate it with a session identifier on your server.
+ */
+class PayPalDataCollector internal constructor(
+    coreConfig: CoreConfig,
+    private val magnesSDK: MagnesSDK,
+    private val uuidHelper: UUIDHelper
+) {
+
+    private val environment = when(coreConfig.environment) {
+        Environment.LIVE -> lib.android.paypal.com.magnessdk.Environment.LIVE
+        Environment.STAGING -> lib.android.paypal.com.magnessdk.Environment.STAGE
+        Environment.SANDBOX -> lib.android.paypal.com.magnessdk.Environment.SANDBOX
+    }
+
+    constructor(coreConfig: CoreConfig) : this(
+        coreConfig,
+        MagnesSDK.getInstance(),
+        UUIDHelper()
+    )
+
+    /**
+     * Gets a Client Metadata ID at the time of payment activity. Once a user initiates a PayPal payment
+     * from their device, PayPal uses the Client Metadata ID to verify that the payment is
+     * originating from a valid, user-consented device and application. This helps reduce fraud and
+     * decrease declines. This method MUST be called prior to initiating a pre-consented payment (a
+     * "future payment") from a mobile device. Pass the result to your server, to include in the
+     * payment request sent to PayPal. Do not otherwise cache or store this value.
+     *
+     * @param context Android Context
+     * @param clientMetadataId The desired pairing ID, trimmed to 32 characters.
+     * @param additionalData Additional data that should be associated with the data collection
+     *
+     * @return clientMetadataId Your server will send this to PayPal
+     */
+    @JvmOverloads
+    fun getClientMetadataId(context: Context, clientMetadataId: String? = null, additionalData: HashMap<String, String>? = null): String {
+        val appContext = context.applicationContext
+        return try {
+            val magnesSettingsBuilder = MagnesSettings.Builder(appContext)
+                .setMagnesSource(MagnesSource.PAYPAL)
+                .disableBeacon(false)
+                .setMagnesEnvironment(environment)
+                .setAppGuid(uuidHelper.getInstallationGUID(context))
+            magnesSDK.setUp(magnesSettingsBuilder.build())
+            val result = magnesSDK.collectAndSubmit(
+                appContext,
+                clientMetadataId,
+                additionalData
+            )
+            result.paypalClientMetaDataId
+        } catch (e: InvalidInputException) {
+            // Either clientMetadataId or appGuid exceeds their character limit
```

Comment:
Should this error be surfaced somehow or do we want to fail silently on data collection?

Reply:
Braintree makes it fail silently. Perhaps the rationale behind that is that data collection is not a super important part of the merchant expierence? We could just make the method throw the exception and thats it.

Reply:
Ah you're right - failing silently is probably the way to go 👍